File: Shared/walls/walls.cpp
#include "walls.hpp"
#include "compress.hpp"

Walls::Wall* Walls::getWall(int x, int y) {
    if(x < 0 || x >= getWidth() || y < 0 || y >= getHeight())
        throw Exception("Wall is accessed out of the bounds! (" + std::to_string(x) + ", " + std::to_string(y) + ")");
    return &walls[y * getWidth() + x];
}

void Walls::create() {
    walls = new Wall[getWidth() * getHeight()];
}

WallType* Walls::getWallType(int x, int y) {
    return getWallTypeById(getWall(x, y)->id);
}

void Walls::setWallTypeSilently(int x, int y, WallType* type) {
    getWall(x, y)->id = type->id;
}

void Walls::setWallType(int x, int y, WallType* type) {
    if(type->id != getWall(x, y)->id) {
        setWallTypeSilently(x, y, type);
        
        WallChangeEvent event(x, y);
        wall_change_event.call(event);
    }
}

int Walls::getWidth() const {
    return blocks->getWidth();
}

int Walls::getHeight() const {
    return blocks->getHeight();
}

std::vector<char> Walls::toSerial() {
    std::vector<char> serial;
    unsigned long iter = 0;
    serial.resize(serial.size() + getWidth() * getHeight());
    Wall* wall = walls;
    for(int i = 0; i < getWidth() * getHeight(); i++) {
        serial[iter++] = (char)wall->id;
        wall++;
    }
    return compress(serial);
}

void Walls::fromSerial(const std::vector<char>& serial) {
    std::vector<char> decompressed = decompress(serial);
    const char* iter = &decompressed[0];
    create();
    Wall* wall = walls;
    for(int i = 0; i < getWidth() * getHeight(); i++) {
        wall->id = *iter++;
        wall++;
    }
}

void Walls::registerNewWallType(WallType* wall_type) {
    wall_type->id = curr_id++;
    wall_types.push_back(wall_type);
}

WallType* Walls::getWallTypeById(int wall_id) {
    return wall_types[wall_id];
}

WallType* Walls::getWallTypeByName(const std::string& name) {
    for(int i = 0; i < wall_types.size(); i++)
        if(wall_types[i]->name == name)
            return wall_types[i];
    return nullptr;
}

int Walls::getNumWallTypes() {
    return (int)wall_types.size();
}

int Walls::getBreakProgress(int x, int y) {
    for(int i = 0; i < breaking_walls.size(); i++)
        if(breaking_walls[i].x == x && breaking_walls[i].y == y)
            return breaking_walls[i].break_progress;
    return 0;
}

void Walls::updateBreakingWalls(int frame_length) {
    for(int i = 0; i < breaking_walls.size(); i++) {
        if(breaking_walls[i].is_breaking) {
            breaking_walls[i].break_progress += frame_length;
            if(breaking_walls[i].break_progress > getWallType(breaking_walls[i].x, breaking_walls[i].y)->break_time)
                breakWall(breaking_walls[i].x, breaking_walls[i].y);
        }
    }
}

int Walls::getBreakStage(int x, int y) {
    return (float)getBreakProgress(x, y) / (float)getWallType(x, y)->break_time * 9.f;
}

void Walls::startBreakingWall(int x, int y) {
    if(x < 0 || x >= getWidth() || y < 0 || y >= getHeight())
        throw Exception("Wall is accessed out of the bounds! (" + std::to_string(x) + ", " + std::to_string(y) + ")");
    
    BreakingWall* breaking_wall = nullptr;
    
    for(int i = 0; i < breaking_walls.size(); i++)
        if(breaking_walls[i].x == x && breaking_walls[i].y == y)
            breaking_wall = &breaking_walls[i];
    
    if(!breaking_wall) {
        BreakingWall new_breaking_wall;
        new_breaking_wall.x = x;
        new_breaking_wall.y = y;
        breaking_walls.push_back(new_breaking_wall);
        breaking_wall = &breaking_walls.back();
    }
    
    breaking_wall->is_breaking = true;
        
    WallStartedBreakingEvent event(x, y);
    wall_started_breaking_event.call(event);
}

void Walls::stopBreakingWall(int x, int y) {
    if(x < 0 || x >= getWidth() || y < 0 || y >= getHeight())
        throw Exception("Wall is accessed out of the bounds! (" + std::to_string(x) + ", " + std::to_string(y) + ")");
    
    for(int i = 0; i < breaking_walls.size(); i++)
        if(breaking_walls[i].x == x && breaking_walls[i].y == y) {
            breaking_walls[i].is_breaking = false;
            WallStoppedBreakingEvent event(x, y);
            wall_stopped_breaking_event.call(event);
        }
}

void Walls::breakWall(int x, int y) {
    WallBreakEvent event(x, y);
    wall_break_event.call(event);
    
    setWallType(x, y, &clear);
}

File: shared/walls/walls.rs
use anyhow::Result;
use serde_derive::{Deserialize, Serialize};

use crate::libraries::container_file;
use crate::libraries::grid::Grid;
use crate::libraries::registry::{Registry, RegistryId};
use crate::shared::blocks::Tool;
use crate::shared::blocks::{Blocks, ToolId};
use crate::shared::walls::{BreakingWall, Wall};

/// `WallId` stores id to a type of wall.
#[derive(Deserialize, Serialize, Clone, Copy, Eq, PartialEq, PartialOrd, Ord, Hash)]
pub struct WallId {
    pub id: i8,
}

impl WallId {
    #[must_use]
    pub const fn undefined() -> Self {
        Self { id: -1 }
    }
}

impl RegistryId for WallId {
    const KIND: &'static str = "wall type";

    fn from_index(index: usize) -> Self {
        Self { id: index as i8 }
    }

    fn index(self) -> Option<usize> {
        usize::try_from(self.id).ok()
    }
}

#[derive(Deserialize, Serialize)]
pub(super) struct WallsData {
    pub(super) walls: Grid<WallId>,
}

impl WallsData {
    pub const fn new() -> Self {
        Self { walls: Grid::new_empty() }
    }
}

pub struct Walls {
    pub(super) walls_data: WallsData,

    pub(super) breaking_walls: Vec<BreakingWall>,
    pub(super) wall_types: Registry<WallId, Wall>,

    pub clear: WallId,
    pub hammer: ToolId,
}

impl Walls {
    pub fn new(blocks: &mut Blocks) -> Self {
        let mut result = Self {
            walls_data: WallsData::new(),

            breaking_walls: Vec::new(),
            wall_types: Registry::new(),

            clear: WallId::undefined(),
            hammer: ToolId::new(),
        };

        let mut clear = Wall::new();
        "clear".clone_into(&mut clear.name);
        result.clear = Self::register_new_wall_type(&mut result.wall_types, clear);

        let mut hammer = Tool::new();
        "hammer".clone_into(&mut hammer.name);
        result.hammer = blocks.register_new_tool_type(hammer);

        result
    }

    /// Creates a map with the given dimensions, every cell holding `WallId::undefined()`.
    ///
    /// **That fill is a trap for anyone who calls this directly**: reading a wall before
    /// setting one gets an id no wall type answers to, and `get_wall_type` errors on it.
    /// Only `create_from_wall_ids` calls this, and it overwrites every cell immediately.
    pub fn create(&mut self, size: (u32, u32)) {
        self.walls_data.walls = Grid::filled(size, WallId::undefined());
    }

    /// Returns the wall id at the given position.
    fn get_wall(&self, x: i32, y: i32) -> Result<WallId> {
        Ok(*self.walls_data.walls.get(x, y)?)
    }

    #[must_use]
    pub const fn get_size(&self) -> (u32, u32) {
        self.walls_data.walls.get_size()
    }

    /// Returns the wall type of the wall at given x and y
    pub fn get_wall_type_at(&self, x: i32, y: i32) -> Result<Wall> {
        self.get_wall_type(self.get_wall(x, y)?)
    }

    /// Returns the wall type with the given id.
    ///
    /// A clone rather than a reference, unlike `Blocks::get_block_type` - the callers hold
    /// a lock while they use it, and handing back a borrow of the guard's contents is a
    /// separate change from this one.
    pub fn get_wall_type(&self, id: WallId) -> Result<Wall> {
        Ok(self.wall_types.get(id)?.clone())
    }

    /// This function sets the wall type on x and y and sends the `WallChangeEvent`.
    pub fn set_wall_type(&mut self, x: i32, y: i32, wall_id: WallId) -> Result<()> {
        let wall = self.get_wall(x, y)?;
        if wall == wall_id {
            return Ok(());
        }

        self.walls_data.walls.set(x, y, wall_id)?;

        Ok(())
    }

    /// Serializes walls for saving
    pub fn serialize(&self) -> Result<Vec<u8>> {
        container_file::pack(&self.walls_data)
    }

    /// Deserializes walls from u8 vector
    pub fn deserialize(&mut self, data: &[u8]) -> Result<()> {
        self.walls_data = container_file::unpack(data)?;

        Ok(())
    }

    /// This function adds a new wall type, but is used internally by mods.
    pub(super) fn register_new_wall_type(wall_types: &mut Registry<WallId, Wall>, wall_type: Wall) -> WallId {
        wall_types.register(wall_type)
    }

    /// Returns a wall id type with the given name
    pub fn get_wall_id_by_name(&self, name: &str) -> Result<WallId> {
        self.wall_types.get_id_by_name(name)
    }

    /// This function creates a world from a 2d vector of wall type ids
    pub fn create_from_wall_ids(&mut self, wall_ids: &[Vec<WallId>]) -> Result<()> {
        self.walls_data.walls = Grid::from_columns(wall_ids)?;
        Ok(())
    }

    /// Returns all wall ids.
    #[must_use]
    pub fn get_all_wall_ids(&self) -> Vec<WallId> {
        self.wall_types.ids()
    }

    /// Returns all breaking walls
    #[must_use]
    pub const fn get_breaking_walls(&self) -> &Vec<BreakingWall> {
        &self.breaking_walls
    }
}

pub struct WallChangeEvent {
    pub x: i32,
    pub y: i32,
}

pub struct WallBreakEvent {
    pub x: i32,
    pub y: i32,
}

pub struct WallStartedBreakingEvent {
    pub x: i32,
    pub y: i32,
}

pub struct WallStoppedBreakingEvent {
    pub x: i32,
    pub y: i32,
}

/// A welcome packet that carries all the information about the world walls
#[derive(Serialize, Deserialize)]
pub struct WallsWelcomePacket {
    pub data: Vec<u8>,
}
